### PR TITLE
RSDK-8250, RSDK-8564, RSDK-8599 - Configurable logging cleanup

### PR DESCRIPTION
**Pull request by dgottlieb**
From @benjirewis: 

RSDK-8250, RSDK-8564, RSDK-8599

This PR "cleans up" some of the configurable logging code. See my GH comments for more details, and feel free to ask questions offline.

Dan's original comment:

> This isn't exactly done, but leaving it in this state while I'm away. As discussed, feel free to leave this alone or to take ownership/merge with config reading changes.

### CODE REVIEW

File: config/reader_test.go [hunk 1]
```diff
@@ -370,103 +368,103 @@ func TestReadTLSFromCache(t *testing.T) {
...
-	test.That(t, logger.GetLevel().String(), test.ShouldEqual, "Info")
-
-	logger, ok = logging.LoggerNamed(componentLoggerName)
-	test.That(t, ok, test.ShouldBeTrue)
-	test.That(t, logger.GetLevel().String(), test.ShouldEqual, "Info")
-
-	test.That(t, logging.DeregisterLogger(serviceLoggerName), test.ShouldBeTrue)
-	test.That(t, logging.DeregisterLogger(componentLoggerName), test.ShouldBeTrue)
-}
+// func TestProcessConfigRegistersLogConfig(t *testing.T) {
```
Comment (dgottlieb) [inline config/reader_test.go:371]: Should probably be re-instantiated. I commented it out so I could freely figure out the "process config" apis. But I ran out of time and wanted to share this now.

I did*  remove the public "DeregisterLogger" API. Nothing used it but this test. I think `registery.deregisterLogger` still exists if we even need it.

File: robot/impl/robot_framesystem_test.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ import (
 func TestFrameSystemConfigWithRemote(t *testing.T) {
 	logger := logging.NewTestLogger(t)
 	// make the remote robots
-	remoteConfig, err := config.Read(context.Background(), rutils.ResolveFile("robot/impl/data/fake.json"), logger.Sublogger("remote"))
+	remoteConfig, err := config.Read(context.Background(), rutils.ResolveFile("robot/impl/data/fake.json"), logger.RootSublogger("remote"))
```
Comment (dgottlieb) [inline robot/impl/robot_framesystem_test.go:33]: I was only running tests in impl to suss out build errors. I suspect there are other _test files that are failing CI and need a similar change.

File: logging/logger_registry.go [hunk 3]
```diff
@@ -37,17 +34,15 @@ func (lr *loggerRegistry) deregisterLogger(name string) bool {
...
 	defer lr.mu.RUnlock()
 	logger, ok = lr.loggers[name]
 	return
 }
 
-func (lr *loggerRegistry) updateLoggerLevelWithCfg(name string) error {
-	lr.mu.RLock()
-	defer lr.mu.RUnlock()
-
+// updateLoggerLevelWithCfg must be called with a read or write lock on `lr.mu`.
```
Comment (dgottlieb) [inline logging/logger_registry.go:44]: Highlighting this. I also want to take an axe to the private API because these private functions are either:
* not called
* only called by other private functions

Making the locking requirements confusing.

But I'm saving that for some later refactor after the behavior changes/requirements are locked down.

File: logging/logger_registry_test.go [hunk 4]
```diff
@@ -13,15 +14,50 @@ var (
...
+func (mem *memAppender) Write(entry zapcore.Entry, field []zapcore.Field) error {
+	mem.entries = append(mem.entries, &entry)
+	return nil
+}
+
+func (mem *memAppender) Sync() error {
+	return nil
+}
+
+func TestRegistryIsolation(t *testing.T) {
```
Comment (dgottlieb) [inline logging/logger_registry_test.go:39]: I started this to show my new code fixed a bug. But then I realized I couldn't even express the bug with the `GetOrNewLogger` free-floating function. So it was pointless. Feel free to delete.

File: logging/logger_registry.go [hunk 5]
```diff
@@ -119,87 +117,40 @@ func (lr *loggerRegistry) updateLoggerRegistry(logConfig []LoggerPatternConfig)
...
 	defer lr.mu.RUnlock()
-	registeredNames := make([]string, 0, len(globalLoggerRegistry.loggers))
+	registeredNames := make([]string, 0, len(lr.loggers))
 	for name := range lr.loggers {
 		registeredNames = append(registeredNames, name)
 	}
 	return registeredNames
 }
 
-func (lr *loggerRegistry) registerConfig(logConfig []LoggerPatternConfig) error {
```
Comment (dgottlieb) [inline logging/logger_registry.go:132]: I didn't see why we'd ever call `updateLoggerRegistry` without setting the `logConfig`. I allowed myself this API change to reduce the number of possible states. Such that I could convince myself we weren't creating problems with weird calls to `updateLoggerRegistry` that intentionally weren't calling `registerConfig`.

Never got to the convincing part. Just seeing this now reminded me to double check.

File: config/config_test.go [hunk 6]
```diff
@@ -167,13 +167,13 @@ func TestConfigWithLogDeclarations(t *testing.T) {
 	// log configure for builtin fake motors would apply for a log level of `warn`. This "overlayed"
 	// log level is not applied at config parsing time.
 	test.That(t, cfg.Components[2].Name, test.ShouldEqual, "right_motor")
-	test.That(t, cfg.Components[2].LogConfiguration.Level, test.ShouldEqual, logging.INFO)
+	test.That(t, cfg.Components[2].LogConfiguration, test.ShouldBeNil)
```
Comment (benjirewis) [inline config/config_test.go:170]: `LogConfiguration` is now an "optional" field (represented by a pointer). We used to use `logging.INFO` as the default level, but it is semantically better to use `nil` to represent _no_ specified configuration.

File: config/diff.go [hunk 7]
```diff
@@ -513,3 +517,16 @@ func diffModule(left, right Module, diff *Diff) bool {
...
 	return true
 }
+
+// diffLogCfg returns true if any part of the log config is different or if any
+// services or components have been updated.
+func diffLogCfg(left, right *Config, servicesDifferent, componentsDifferent bool) bool {
+	if !reflect.DeepEqual(left.LogConfig, right.LogConfig) {
+		return true
+	}
+	// If there was any change in services or components; attempt to update logger levels.
```
Comment (benjirewis) [inline config/diff.go:527]: I've added some basic log diffing here. It is lazy in the sense that I don't _actually_ look through `LogConfiguration` fields on services and components to determine if anything has changed (I treat any change in services or components as a potential change in logging behavior). I was lazy because I'm really only using the returned boolean here to avoid calling `s.registry.Update` on _every_ config update, and output a `Detected potential log configuration changes; updating logger levels` log.

File: config/reader.go [hunk 8]
```diff
@@ -610,16 +610,6 @@ func processConfig(unprocessedConfig *Config, fromCloud bool, logger logging.Log
 		}
 	}
 
-	// pattern match configurations and resource log level configurations are both necessary for the logger registry
```
Comment (benjirewis) [inline config/reader.go:613]: All logic removed from this file has moved to `web/server/entrypoint.go`. As discussed offline, we do not want config processing to mutate the registry because a) we want fewer side-effects from these functions and b) we would have to pass the registry object down through these functions.

File: config/reader.go [hunk 9]
```diff
@@ -739,7 +708,7 @@ func CreateNewGRPCClient(ctx context.Context, cloudCfg *Cloud, logger logging.Lo
 		dialOpts = append(dialOpts, rpc.WithInsecure())
 	}
 
-	return rpc.DialDirectGRPC(ctx, u.Host, logging.GetOrNewLogger("rdk.networking"), dialOpts...)
```
Comment (benjirewis) [inline config/reader.go:742]: `GetOrNewLogger` is gone. In this case, it would have gotten a logger named `rdk.networking` from the registry and created it if it didn't exist. That logic is now handled automatically by `Sublogger`.

File: config/reader_test.go [hunk 10]
```diff
@@ -369,104 +367,3 @@ func TestReadTLSFromCache(t *testing.T) {
 		test.That(t, err, test.ShouldBeNil)
 	})
 }
-
-func TestProcessConfigRegistersLogConfig(t *testing.T) {
```
Comment (benjirewis) [inline config/reader_test.go:373]: I will replace this test with a higher level one (something in the `web/server` package) once I get feedback on my other changes.

File: logging/log_config.go [hunk 11]
```diff
@@ -25,7 +25,8 @@ const (
 	validResourceSubTypePattern = validNamespacePattern
 	validModelNamePattern       = validNamespacePattern
 	validTypeSubsectionPattern  = `(` + validResourceTypePattern + `:` + validResourceSubTypePattern + `|remote:)`
-	validResourcePattern        = `^rdk.` + validNamespacePattern + `:` + validTypeSubsectionPattern + `\/` + validModelNamePattern + `$`
+	validResourcePattern        = `^rdk.resource_manager.` + validNamespacePattern + `:` + validTypeSubsectionPattern + `\/` +
+		validModelNamePattern + `$`
```
Comment (benjirewis) [inline logging/log_config.go:29]: This regex was slightly incorrect (it was correct at an earlier point). "Resource patterns" start with `rdk.resource_manager`, not just `rdk`.

File: logging/log_config.go [hunk 12]
```diff
@@ -43,7 +44,7 @@ func buildRegexFromPattern(pattern string) string {
 	for _, ch := range pattern {
 		switch ch {
 		case '*':
-			matcher.WriteString(validLoggerSections)
+			matcher.WriteString(`.*`)
```
Comment (benjirewis) [inline logging/log_config.go:47]: I don't understand why this was `validLoggerSections`; `*` in a `log` pattern should be translated to `.*`. The interns seemed to agree before they left that this was a mistake.

File: logging/logger_registry.go
Comment (benjirewis) [inline logging/logger_registry.go:1]: Lots of changes in this file; hopefully they are somewhat self-explanatory. Generally, we've simplified the API a bit, gotten rid of the `globalLoggerRegistry` and its associated exported functions.

File: logging/logger_registry_test.go [hunk 13]
```diff
@@ -13,103 +12,73 @@ var (
 	l3 = NewLogger("logger-3")
 )
 
-func mockRegistry() *loggerRegistry {
-	manager := newLoggerRegistry()
```
Comment (benjirewis) [inline logging/logger_registry_test.go:17]: `manager` -> `registry`. In an old version of configurable logging, registries were called "managers". No longer.

File: logging/logger_registry_test.go [hunk 14]
```diff
@@ -13,103 +12,73 @@ var (
...
 	test.That(t, ok, test.ShouldBeTrue)
 	test.That(t, logger1, test.ShouldEqual, l1)
 
-	sublogger2, ok := manager.loggerNamed("sublogger-2")
+	sublogger2, ok := registry.loggerNamed("sublogger-2")
 	test.That(t, ok, test.ShouldBeFalse)
 	test.That(t, sublogger2, test.ShouldBeNil)
 }
 
-func TestLoggersRegisteredOnCreation(t *testing.T) {
```
Comment (benjirewis) [inline logging/logger_registry_test.go:56]: Not a concept without global registry.

File: logging/logging.go [hunk 15]
```diff
@@ -65,23 +65,41 @@ func NewLogger(name string) Logger {
...
 	}
 
-	RegisterLogger(name, logger)
+	logger.registry.registerLogger(name, logger)
 	return logger
 }
 
+// NewLoggerWithRegistry is the same as NewLogger but also returns the
+// associated Registry.
+func NewLoggerWithRegistry(name string) (Logger, *Registry) {
```
Comment (benjirewis) [inline logging/logging.go:78]: This is the new method we discussed offline. Used in entrypoint code to create the highest level `rdk` logger and update its associated registry whenever a new config comes in.

File: logging/net_appender.go [hunk 16]
```diff
@@ -35,20 +35,32 @@ type CloudConfig struct {
 // NewNetAppender creates a NetAppender to send log events to the app backend. NetAppenders ought to
 // be `Close`d prior to shutdown to flush remaining logs.
 // Pass `nil` for `conn` if you want this to create its own connection.
-func NewNetAppender(config *CloudConfig, conn rpc.ClientConn, sharedConn bool) (*NetAppender, error) {
-	return newNetAppender(config, conn, sharedConn, true)
+func NewNetAppender(
+	config *CloudConfig,
+	conn rpc.ClientConn,
+	sharedConn bool,
+	loggerWithoutNet Logger,
```
Comment (benjirewis) [inline logging/net_appender.go:42]: I passed in a real `loggerWithoutNet` to net appenders, so we could have it be a real `Sublogger` of the `rdk.networking` logger.

File: resource/config_test.go [hunk 17]
```diff
@@ -11,9 +11,22 @@ import (
 	"go.viam.com/rdk/components/sensor"
 	"go.viam.com/rdk/resource"
 	"go.viam.com/rdk/services/motion"
-	"go.viam.com/rdk/testutils"
 )
 
+// FakeConvertedAttributes is a helper for testing if validation works.
+type FakeConvertedAttributes struct {
```
Comment (benjirewis) [inline resource/config_test.go:17]: [drive-by] Only usage of this struct was in this file. @dgottlieb moved it here and removed `testutils/config_utils.go`.

File: resource/graph_node.go [hunk 18]
```diff
@@ -169,9 +169,8 @@ func (w *GraphNode) TransitionedAt() time.Time {
 }
 
 // InitializeLogger initializes the logger object associated with this resource node.
-func (w *GraphNode) InitializeLogger(parent logging.Logger, subname string, level logging.Level) {
+func (w *GraphNode) InitializeLogger(parent logging.Logger, subname string) {
 	logger := parent.Sublogger(subname)
-	logger.SetLevel(level)
```
Comment (benjirewis) [inline resource/graph_node.go:174]: Levels handled at a much higher level now (entrypoint code). Do not need this nor the `SetLogLevel` method below.

File: resource/graph_node.go [hunk 19]
```diff
@@ -539,15 +529,6 @@ func (w *GraphNode) ResourceStatus() Status {
 	return w.resourceStatus()
 }
 
-func (w *GraphNode) getLoggerOrGlobal() logging.Logger {
```
Comment (benjirewis) [inline resource/graph_node.go:542]: @maximpertsov I removed this and nothing broke so will take it and its usage out unless you disagree.

File: robot/impl/resource_manager_modular_test.go [hunk 20]
```diff
@@ -479,156 +476,6 @@ func (m *dummyModMan) Close(ctx context.Context) error {
 	return nil
 }
 
-func TestDynamicModularComponentLogging(t *testing.T) {
```
Comment (benjirewis) [inline robot/impl/resource_manager_modular_test.go:482]: I will replace this test and the one below with higher level ones (something in the `web/server` package) once I get feedback on my other changes.

File: web/server/entrypoint.go [hunk 21]
```diff
@@ -216,6 +209,37 @@ func (s *robotServer) createWebOptions(cfg *config.Config) (weboptions.Options,
 	return options, nil
 }
 
+func (s *robotServer) updateLoggerRegistry(cfg *config.Config) {
+	var combinedLogCfg []logging.LoggerPatternConfig
+	if cfg.LogConfig != nil {
+		combinedLogCfg = append(combinedLogCfg, cfg.LogConfig...)
+	}
+
+	for _, serv := range cfg.Services {
```
Comment (benjirewis) [inline web/server/entrypoint.go:218]: We still rely on ordering here. `LogConfig` processed before `LogConfiguration`s means `LogConfiguration`s take precedence over `LogConfig`.

File: web/server/entrypoint.go [hunk 22]
```diff
@@ -410,6 +439,12 @@ func (s *robotServer) serveWeb(ctx context.Context, cfg *config.Config) (err err
 					}
 				}
 
+				// Update logger registry if log patterns may have changed.
+				if !diff.LogEqual {
+					s.logger.Debug("Detected potential changes to log patterns; updating logger levels")
```
Comment (benjirewis) [inline web/server/entrypoint.go:444]: Could bump to `Info` but may be overly noisy.

Comment (cheukt) [reply web/server/entrypoint.go:444]: I'd be interested if we could debug log all the patterns easily, but ya, debug is probably the right level

File: robot/impl/resource_manager_modular_test.go (hunk 20 above)
Comment (cheukt) [reply robot/impl/resource_manager_modular_test.go:482]: curious about why we have to move the test

**Review by cheukt — COMMENTED** [review]
took a pass, I think the changes all look good! only have 1 real comment around where a helper could go. can you also add the tickets the pr links to in the title?

File: resource/graph_node.go (hunk 19 above)
Comment (maximpertsov) [reply resource/graph_node.go:542]: yep! it's possible that we never needed to guard against a `nil` logger for this code path in the first place. I'm curious if we can remove the remaining if `w.logger != nil` guards in this file. But no need to do that yet, this change is large enough as is.

File: web/server/entrypoint.go (hunk 21 above)
Comment (maximpertsov) [reply web/server/entrypoint.go:218]: @cheukt you mean a helper function that just returns a `[]logging.LoggerPatternConfig`? I like that idea. `logging` feels like the right place to have such a helper function.

Comment (cheukt) [reply web/server/entrypoint.go:218]: either one that returns `[]logging.LoggerPatternConfig` or one that acts upon the registry and updates it as well

File: web/server/entrypoint.go (hunk 22 above)
Comment (benjirewis) [reply web/server/entrypoint.go:444]: I agree; given "size" of this PR (number of things I'm trying to cover), will punt to https://viam.atlassian.net/browse/RSDK-8712.

File: web/server/entrypoint.go (hunk 21 above)
Comment (benjirewis) [reply web/server/entrypoint.go:218]: Good idea; moved it to be a helper in `config`. I agree that `logging` would be a better package, but `config` already depends on `logging`, so there would be an import cycle.

File: robot/impl/resource_manager_modular_test.go (hunk 20 above)
Comment (benjirewis) [reply robot/impl/resource_manager_modular_test.go:482]: This test relies on changes to `LogConfiguration` for a resource actually modifying the logger registry. Modifications to the logger registry now only happen in `web/server/entrypoint.go`. If you're asking why I need to move the test to `web/server`, though, I suspect I'll have to do that to access some private API from `web/server` (particularly `runServer`).

File: logging/log_config.go (hunk 11 above)
Comment (benjirewis) [reply logging/log_config.go:29]: Documented all the regexes with examples.

File: web/server/entrypoint.go (hunk 21 above)
Comment (maximpertsov) [reply web/server/entrypoint.go]: > so there would be an import cycle

the function could take in just the `resource.Config` and `logging.Config` fields to avoid the import cycle

Comment (benjirewis) [reply web/server/entrypoint.go]: > take in just the `resource.Config`

Unfortunately `logging` depending on `resource` would also be an import cycle. Most of our packages depend on `logging`, so helpers _in_ `logging` that reference other packages will usually result in cycles.

**Comment by benjirewis** [comment]
Will add testing if others do not have big comments.

**Review by maximpertsov — APPROVED** [review]

File: robot/web/web_c.go [hunk 23]
```diff
@@ -334,7 +334,7 @@ func (svc *webService) closeStreamServer() {
 	}
 }
 
-func (svc *webService) initStreamServer(ctx context.Context, options *weboptions.Options) error {
+func (svc *webService) initStreamServer(ctx context.Context, options *weboptions.Options, logger logging.Logger) error {
```
Comment (cheukt) [inline robot/web/web_c.go:337]: is this addition needed?

**Review by cheukt — COMMENTED** [review]
one comment, but otherwise looks good!

File: robot/web/web_c.go (hunk 23 above)
Comment (benjirewis) [reply robot/web/web_c.go:337]: Nope! Good eye. I think these CGO-interacting files do not get linted for some reason.

**Review by cheukt — APPROVED** [review]
LGTM
